graph/src/schema/input_schema.rs

```diff
@@ -0,0 +1,357 @@
+use std::collections::{BTreeMap, HashSet};
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::{anyhow, Error};
+use store::Entity;
+
+use crate::cheap_clone::CheapClone;
+use crate::components::store::{EntityKey, EntityType, LoadRelatedRequest};
+use crate::data::graphql::ext::DirectiveFinder;
+use crate::data::graphql::{DirectiveExt, DocumentExt, ObjectTypeExt, TypeExt, ValueExt};
+use crate::data::store::{self, scalar, IntoEntityIterator, TryIntoEntityIterator};
+use crate::prelude::q::Value;
+use crate::prelude::{s, DeploymentHash};
+use crate::schema::api_schema;
+use crate::util::intern::AtomPool;
+
+use super::fulltext::FulltextDefinition;
+use super::{ApiSchema, Schema, SchemaValidationError};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct InputSchema {
+    inner: Arc<Inner>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Inner {
+    schema: Schema,
+    immutable_types: HashSet<EntityType>,
+    pool: Arc<AtomPool>,
+}
+
+impl std::ops::Deref for InputSchema {
+    type Target = Inner;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl CheapClone for InputSchema {
+    fn cheap_clone(&self) -> Self {
+        InputSchema {
+            inner: self.inner.cheap_clone(),
+        }
+    }
+}
+
+impl InputSchema {
+    fn create(schema: Schema) -> Self {
+        let immutable_types = HashSet::from_iter(
+            schema
+                .document
+                .get_object_type_definitions()
+                .into_iter()
+                .filter(|obj_type| obj_type.is_immutable())
+                .map(Into::into),
+        );
+
+        let pool = Arc::new(atom_pool(&schema.document));
+
+        Self {
+            inner: Arc::new(Inner {
+                schema,
+                immutable_types,
+                pool,
+            }),
+        }
+    }
+
+    pub fn new(id: DeploymentHash, document: s::Document) -> Result<Self, SchemaValidationError> {
+        let schema = Schema::new(id, document)?;
+        Ok(Self::create(schema))
+    }
+
+    pub fn parse(raw: &str, id: DeploymentHash) -> Result<Self, Error> {
+        let schema = Schema::parse(raw, id)?;
+
+        Ok(Self::create(schema))
+    }
+}
+
+impl Inner {
+    pub fn api_schema(&self) -> Result<ApiSchema, anyhow::Error> {
```

Comment:
could you also add some comments to these other exported fns ?

Reply:
A lot of this is just existing code moved here from other places - for better or worse, a lot of that doesn't have comments, and commenting it all would be a pretty big undertaking. I'll add some comments, but can't do that for everything here.